### PR TITLE
Simplify parsing of errorbar input.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -3108,41 +3108,27 @@ def xywhere(xs, ys, mask):
...
-                    high = [thisx + thiserr for thisx, thiserr
-                            in cbook.safezip(data, b)]
-                    return low, high
-            # Check if xerr is scalar or symmetric. Asymmetric is handled
-            # above. This prevents Nx2 arrays from accidentally
-            # being accepted, when the user meant the 2xN transpose.
-            # special case for empty lists
-            if len(err) > 1:
-                fe = cbook.safe_first_element(err)
-                if len(err) != len(data) or np.size(fe) > 1:
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: the check that len(err) == len(data) is taken care of by safezip, so can be dropped.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -3108,41 +3108,27 @@ def xywhere(xs, ys, mask):
...
-            low = [thisx - thiserr for thisx, thiserr
-                   in cbook.safezip(data, err)]
-            high = [thisx + thiserr for thisx, thiserr
-                    in cbook.safezip(data, err)]
+                a = b = err  # Symmetric error: 1D iterable.
+            # This could just be `np.ndim(a) > 1 and np.ndim(b) > 1`, except
+            # for the (undocumented, but tested) support for (n, 1) arrays.
+            ash, bsh = map(np.shape, [a, b])
+            if (len(ash) > 2 and not (len(ash) == 2 and ash[1] == 1)
+                    or len(bsh) > 2 and not (len(bsh) == 2 and bsh[1] == 1)):
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]: I think the expression is wrong. The first part is already False for `len(ashape) == 2` so the rhs of the `and` is never evaluated in that case. Should be 

~~~
if (len(ashape) > 2 or (len(ashape) == 2 and ashape[1] != 1)
~~~

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -3108,41 +3108,27 @@ def xywhere(xs, ys, mask):
...
+                raise ValueError(
+                    "err must be a scalar or a 1D or (2, n) array-like")
+            # Using list comprehensions rather than arrays to preserve units.
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]: Would it be worth it to special-case if v and e are arrays to speed up the calculation? Probably not.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: duh, indeed

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -3108,41 +3108,27 @@ def xywhere(xs, ys, mask):
...
-                                     "or 2xN array-like ]")
-            # using list comps rather than arrays to preserve units
-            low = [thisx - thiserr for thisx, thiserr
-                   in cbook.safezip(data, err)]
-            high = [thisx + thiserr for thisx, thiserr
-                    in cbook.safezip(data, err)]
+                a = b = err  # Symmetric error: 1D iterable.
+            # This could just be `np.ndim(a) > 1 and np.ndim(b) > 1`, except
+            # for the (undocumented, but tested) support for (n, 1) arrays.
+            ash, bsh = map(np.shape, [a, b])
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: Actually it doesn't (after the fix below), but one can write `a_sh` now...

**Comment by anntzer** [comment]
thanks, handled

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -3108,41 +3108,28 @@ def xywhere(xs, ys, mask):
...
+            # This could just be `np.ndim(a) > 1 and np.ndim(b) > 1`, except
+            # for the (undocumented, but tested) support for (n, 1) arrays.
+            a_sh = np.shape(a)
+            b_sh = np.shape(b)
+            if (len(a_sh) > 2 or (len(a_sh) == 2 and a_sh[1] != 1)
+                    or len(b_sh) > 2 or (len(b_sh) == 2 and b_sh[1] != 1)):
+                raise ValueError(
+                    "err must be a scalar or a 1D or (2, n) array-like")
+            # Using list comprehensions rather than arrays to preserve units.
+            low = [v - e for v, e in cbook.safezip(data, a)]
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: What units has this been tested with?  This is basically the same issue as `bar` in #12903 so we should make sure unit handling is the same.  In particular if v is a datetime, and e a timedelta, I assume this works?  What about for `pint` units?  And are we sure this works w/ pandas?  Not that I think #12903 checked all those boxes, but I think we should be thinking about how to uniformly handle this case and `bar`.  ie. we now have `_convert_dx`, and maybe errorbar should use it as well.

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: Do you want to first make a PR letting errorbar use _convert_dx?  I can rebase this one onto yours after it's merged.

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: I agree they are different and maybe using the same helper doesn't make sense.  (I actually think defaulting `width=0.8` in bar is ambiguous and a mistake, but...).  

Tests with datetime-like obects would be helpful.  Not sure categoricals need to be tested (when would the error be `a +/- f`? )  

And we should decide if we should add `pint` to the tests.

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: tbh I don't particularly care about adding tests for this PR specifically.

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: actually I guess thats fair since you didn't change the algorithm...

**Pull request merged**